src/kernel/bitcoinkernel.cpp

```diff
@@ -1,11 +1,1135 @@
+extern const std::function<std::string(const char*)> G_TRANSLATION_FUN{nullptr};
+
+static const kernel::Context kernel_context_static{};
+
+namespace {
+
+/** Check that all specified flags are part of the libbitcoinkernel interface. */
+bool verify_flags(unsigned int flags)
+{
+    return (flags & ~(kernel_SCRIPT_FLAGS_VERIFY_ALL)) == 0;
+}
+
+bool is_valid_flag_combination(unsigned int flags)
+{
+    if (flags & SCRIPT_VERIFY_CLEANSTACK && ~flags & (SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_WITNESS)) return false;
+    if (flags & SCRIPT_VERIFY_WITNESS && ~flags & SCRIPT_VERIFY_P2SH) return false;
+    return true;
+}
+
+std::string log_level_to_string(const kernel_LogLevel level)
+{
+    switch (level) {
+    case kernel_LogLevel::kernel_LOG_INFO: {
+        return "info";
+    }
+    case kernel_LogLevel::kernel_LOG_DEBUG: {
+        return "debug";
+    }
+    case kernel_LogLevel::kernel_LOG_TRACE: {
+        return "trace";
+    }
+    } // no default case, so the compiler can warn about missing cases
+    assert(false);
+}
+
+std::string log_category_to_string(const kernel_LogCategory category)
+{
+    switch (category) {
+    case kernel_LogCategory::kernel_LOG_BENCH: {
+        return "bench";
+    }
+    case kernel_LogCategory::kernel_LOG_BLOCKSTORAGE: {
+        return "blockstorage";
+    }
+    case kernel_LogCategory::kernel_LOG_COINDB: {
+        return "coindb";
+    }
+    case kernel_LogCategory::kernel_LOG_LEVELDB: {
+        return "leveldb";
+    }
+    case kernel_LogCategory::kernel_LOG_LOCK: {
+        return "lock";
+    }
+    case kernel_LogCategory::kernel_LOG_MEMPOOL: {
+        return "mempool";
+    }
+    case kernel_LogCategory::kernel_LOG_PRUNE: {
+        return "prune";
+    }
+    case kernel_LogCategory::kernel_LOG_RAND: {
+        return "rand";
+    }
+    case kernel_LogCategory::kernel_LOG_REINDEX: {
+        return "reindex";
+    }
+    case kernel_LogCategory::kernel_LOG_VALIDATION: {
+        return "validation";
+    }
+    case kernel_LogCategory::kernel_LOG_KERNEL: {
+        return "kernel";
+    }
+    case kernel_LogCategory::kernel_LOG_ALL: {
+        return "all";
+    }
+    } // no default case, so the compiler can warn about missing cases
+    assert(false);
+}
+
+kernel_SynchronizationState cast_state(SynchronizationState state)
+{
+    switch (state) {
+    case SynchronizationState::INIT_REINDEX:
+        return kernel_SynchronizationState::kernel_INIT_REINDEX;
+    case SynchronizationState::INIT_DOWNLOAD:
+        return kernel_SynchronizationState::kernel_INIT_DOWNLOAD;
+    case SynchronizationState::POST_INIT:
+        return kernel_SynchronizationState::kernel_POST_INIT;
+    } // no default case, so the compiler can warn about missing cases
+    assert(false);
+}
+
+kernel_Warning cast_kernel_warning(kernel::Warning warning)
+{
+    switch (warning) {
+    case kernel::Warning::UNKNOWN_NEW_RULES_ACTIVATED:
+        return kernel_Warning::kernel_LARGE_WORK_INVALID_CHAIN;
```

Comment:
Isn't this supposed to return `kernel_UNKNOWN_NEW_RULES_ACTIVATED`?

```suggestion
        return kernel_Warning::kernel_UNKNOWN_NEW_RULES_ACTIVATED;
```

Reply:
Indeed, I think I mixed this up at some point during a rebase. Thank!